### ISSUE TITLE
qpdf compress/uncompress inconsistenence

### ISSUE THREAD

**Issue by wangf1978** [qpdf/qpdf#271]
I have some PDF files, after uncompress stream-data and then compress stream-data again w/o any changes, the file size become much bigger, here is command lines:
```
qpdf --stream-data=uncompress 6059954.pdf 6059954_u.pdf
qpdf --stream-data=compress 6059954_u.pdf 6059954_c.pdf

6059954.pdf: 2890KB
6059954_c.pdf: 3289KB
```
Is it possible to fix it?

In Acrobat Reader Pro, I save 6059954_c.pdf again, the size become the similar size with the original file.

**Comment by wangf1978** [comment]
[6059954.pdf](https://github.com/qpdf/qpdf/files/2741707/6059954.pdf)

**Comment by jberkenbilt** [comment]
Thanks for including the test file. My guess would be that there is some stream that is compressed in a way that qpdf doesn't replicate by default...maybe it uses decode parameters that make the compression much more efficient or something. To investigate this, I would have to look at the sizes stream by stream.

What is your use case and workflow for this? Maybe I can suggest an alternative workflow that is not prone to the same problem. When I get a chance, I will take a look at the specific file and see if I can determine exactly what the problem is. There are a few features on the qpdf roadmap that will allow replacement of individual streams. This might make it possible to mitigate the effect of this issue, though it would probably remain somewhat complex to do.

**Comment by jberkenbilt** [comment]
If you're using C++ code in the library, you can just manipulate the streams you actually need to modify. If you are using some external program to manipulate the streams in a PDF that is the output of qpdf, I could suggest a workaround. Also, are you using/aware of qdf mode and fix-qdf? These are discussed in the docs (http://qpdf.sourceforge.net/files/qpdf-manual.html#ref.qdf). You could do something like the following:
* use qpdf to generate an output with all the streams uncompressed
* use qpdf to generate a second output with stream compression preserved
* do the replacement of only the streams you want to manipulate in the first file (uncompressed)
* recompress the first file
* in the second file (compression preserved), replace the raw stream data of only the modified streams with the result of the recompression of the modified file
If you use qdf mode and fix-qdf, it will take care of recomputing stream lengths for you. If you know the object ID and generation of the streams you are replacing, you can also use qpdf `--show-object` with `--raw-stream-data` or `--filtered-stream-data` to get data from individual streams. Just bear in mind that the object numbers are modified by qpdf when it writes out a new file, so you'll have to be mindful of whether you're working on original or updated object numbers. qpdf's generation of object numbers at the time of output is deterministic such that object numbers should be preserved if you rewrite output previously created by qpdf and you haven't changed anything about the object or reference structure. While qpdf's test suite currently assumes this functionality, it may not be the best idea to actually rely on that. Though I doubt I would change it...I'm very careful about changing qpdf's output because I don't want to gratuitously break other people's test code....

**Comment by wangf1978** [comment]
Thanks for your nice solution.

I am using C++ to modify the PDF uncompressed by qpdf, and I don't want to change the PDF structure, and reconstruct xref and so on, so all changes are in-place changes, stream length is still kept.
I am afraid the solution does not work for me, because there exists page header in almost every page content stream object, and the main data lies at the page content stream object, almost few unmodified stream from preserved PDF can be shiftted to the new compressed PDF file, right?

BTW:
Is there any option to remove the comment code in output pdf, for example:
%% Original object ID
%% Contents for page

If so, some space may be reduced, or when compressing the pdf again, these comments will be removed?

**Comment by wangf1978** [comment]
Thanks for your nice solution.

I am using C++ to modify the PDF uncompressed by qpdf, and I don't want to change the PDF structure, and reconstruct xref and so on, so all changes are in-place changes, stream length is still kept.
I am afraid the solution does not work for me, because there exists page header in almost every page content stream object, and the main data lies at the page content stream object, almost few unmodified stream from preserved PDF can be shiftted to the new compressed PDF file, right?

BTW:
Is there any option to remove the comment code in output pdf, for example:
%% Original object ID
%% Contents for page

If so, some space may be reduced, or when compressing the pdf again, these comments will be removed?

**Comment by jberkenbilt** [comment]
Q1: those comments are removed when the file is recompressed.
Q2: nothing out of the box. You could do it with various stream data providers or replaceStreamData with memory buffers, but it would require a lot of code, a lot of RAM, or both.
You may also try --object-streams=generate, which will enable the non-stream objects to be compressed. That can save quite a bit of space in some instances. There have been requests to support a more efficient zlib implementation in qpdf, but I have not prioritized that work partly because it will break lots of tests and partly because it adds additional dependencies, and one of the best features of qpdf is that it has extremely few additional dependencies. However, it has been reported to me that you can build your own qpdf and link it with an alternative zlib implementation to get better compression.

**Comment by wangf1978** [comment]
Thanks, let me study how to link with an alternative zlib implementation to get better compression.

**Comment by wangf1978** [comment]
BTW:
--object-streams=generate can get better compression for me, although the output file is still bigger than the original file, but it is much better, I closed this issue.

**Comment by jberkenbilt** [comment]
I've made a note to myself to analyze this in spite of the issue being closed. If I find anything, I'll comment here. Thanks.

**Comment by jberkenbilt** [comment]
There's actually a bug here. If you don't use any `--stream-data` flag at all, it compresses correctly. It seems that when I added the `--compress-streams` and `--decode-level` options, I messed up on the conditions under which object streams and xref streams are compressed. I have fixed this for 8.4.0. Thanks for reporting this and uncovering a subtle bug.
